File: 0101_0200/0185/0185.py
with open("input.txt", "r") as f, open("output.txt", "w") as q:
    n, k = (int(x) for x in f.readline().split())
    m = [[0, set()] for x in range(n + 1)]
    s = [int(x) for x in f.read().split()]
    i = 0
    while s[i] != 0:
        a, b = s[i] if m[s[i]][0] == 0 else m[s[i]][0], s[i + 1]
        i += 2
        for x in m[b][1]:
            m[x][0] = a
        m[b][0] = a
        m[a][1].add(b)
        m[a][1] = set.union(m[a][1], m[b][1])
        set.clear(m[b][1])
    q.write("Yes" if len(m[k][1]) == n - 1 else "No")

File: 0101_0200/0185/0185.c
#include <stdio.h>
#include <stdlib.h>

int check(char *m, int n, int k, char *s)
{
	int r = 0;

	for (int i = 0; i < n; ++i)
		if (m[k * n + i] && !s[i] && (s[i] = 1))
			r += 1 + check(m, n, i, s);
	return r;
}

int main()
{
	FILE *f = fopen("input.txt", "r"), *q = fopen("output.txt", "w");
	int n, k, x, y;
	char *s, *m;

	fscanf(f, "%d%d", &n, &k);
	m = (char*)calloc(n * n, 1);
	s = (char*)calloc(n, 1);
	while (fscanf(f, "%d%d", &x, &y) > 1)
		m[(x - 1) * n + y - 1] = 1;
	fprintf(q, check(m, n, k - 1, s) == n - 1 ? "Yes" : "No");
	return 0;

}
